### PR TITLE
SASS-3788 - Single Amount Is Not Pre-populated On Page: 'Your pension taken in another way'

### PR DESCRIPTION
**Pull request by srinivaspach**
# Checklist

Jira subtask: SASS-3791

## Developer: Before Asking Others To Review Your Code

Have you ...

- [x]  Had a QA verify that all acceptance criteria have been met, with comment(s) in the ticket? 
- [x]  Hosted a group review of your implementation?
- [x]  Rebased against `main`? 
- [x]  Ensured that your commit(s) contain the ticket key?
- [x]  Ensured that your branch is named after the ticket key?
- [x]  Successfully run the unit tests?
- [x]  Successfully run the integration tests?
- [ ]  Successfully run the journey tests? If not, please add a comment explaining why it wasn't necessary.
- [x]  Ensured that the PR job has passed.

## Reviewer: Before Reviewing

Have you ...

- [x]  Pulled the branch down?
- [x]  Assigned yourself to this PR?
- [x]  Moved the Jira *subtask* to “in review”?
- [x]  Ensured that it is rebased against `main`?

## Developer: After Approval

Have you ...

- [ ]  Rebased against `main`?
- [ ]  Moved the Jira *subtask* to “in pipeline”?

## Developer: After Merging

Have you ...

- [ ]  Deleted the branch?
- [ ]  Scheduled a demo? If not, please add a comment.

### CODE REVIEW

**Comment by srinivaspach** [comment]
Successfully run the journey tests? If not, please add a comment explaining why it wasn't necessary. - Journey tests not needed because I did not add any new journey

**Comment by erasmos-hmrc** [comment]
> Successfully run the journey tests? If not, please add a comment explaining why it wasn't necessary.

Anytime we touch production code, we run the journey tests. No exception: not even for me :)

File: app/models/pension/charges/LifetimeAllowance.scala [hunk 1]
```diff
@@ -24,24 +24,27 @@ import utils.DecryptorInstances.bigDecimalDecryptor
 import utils.EncryptableSyntax.EncryptableOps
 import utils.EncryptorInstances.bigDecimalEncryptor
 
-case class LifetimeAllowance(amount: BigDecimal, taxPaid: BigDecimal) {
+case class LifetimeAllowance(
```
Comment (erasmos-hmrc) [inline app/models/pension/charges/LifetimeAllowance.scala:27]: As recommended, could we have both values default to 'None'. That way, we need only construct with LifetimeAllowance().

Also the formatting could be improved.

File: app/models/pension/charges/LifetimeAllowance.scala [hunk 2]
```diff
@@ -24,24 +24,27 @@ import utils.DecryptorInstances.bigDecimalDecryptor
...
+      taxPaid = taxPaid.map(_.encrypted)
     )
 }
 
 object LifetimeAllowance {
   implicit val format: OFormat[LifetimeAllowance] = Json.format[LifetimeAllowance]
 }
 
-case class EncryptedLifetimeAllowance(amount: EncryptedValue, taxPaid: EncryptedValue) {
+case class EncryptedLifetimeAllowance(amount: Option[EncryptedValue], taxPaid: Option[EncryptedValue]) {
```
Comment (erasmos-hmrc) [inline app/models/pension/charges/LifetimeAllowance.scala:43]: Could you have both values default to 'None'.

File: it/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountControllerISpec.scala [hunk 3]
```diff
@@ -259,6 +259,47 @@ class PensionTakenAnotherWayAmountControllerISpec extends IntegrationTest with B
...
+          textOnPageCheck(taxPaidParagraph, taxPaidParagraphSelector)
+          textOnPageCheck(hintText, amount2hintTextSelector, "for amount 2")
+          inputFieldValueCheck(amount1InputName, amount1inputSelector, newAmount.toString)
+          inputFieldValueCheck(amount2InputName, amount2inputSelector, "")
+          buttonCheck(buttonText, continueButtonSelector)
+          formPostLinkCheck(pensionTakenAnotherWayAmountUrl(taxYearEOY), formSelector)
+          welshToggleCheck(user.isWelsh)
+        }
+
+        //"render pension taken another way page with None value for before tax and value for tax paid"
```
Comment (erasmos-hmrc) [inline it/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountControllerISpec.scala:302]: Did you intend to add this scenario?

File: it/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountControllerISpec.scala [hunk 4]
```diff
@@ -19,7 +19,7 @@ package controllers.pensions.lifetimeAllowance
 import builders.PensionLifetimeAllowanceViewModelBuilder.{aPensionLifetimeAllowanceViewModel, aPensionLifetimeAllowancesEmptyViewModel}
 import builders.PensionsUserDataBuilder.{aPensionsUserData, pensionsUserDataWithLifetimeAllowance}
 import builders.UserBuilder.aUserRequest
-import forms.TupleAmountForm
+import forms.{OptionalTupleAmountForm, TupleAmountForm}
```
Comment (erasmos-hmrc) [inline it/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountControllerISpec.scala:22]: TupleAmountForm is unused.

File: app/models/pension/AllPensionsData.scala [hunk 5]
```diff
@@ -75,7 +75,7 @@ object AllPensionsData {
         pensionAsLumpSumQuestion = prior.pensionCharges.flatMap(_.pensionSavingsTaxCharges).map(_.lumpSumBenefitTakenInExcessOfLifetimeAllowance.isDefined),
         pensionAsLumpSum = prior.pensionCharges.flatMap(_.pensionSavingsTaxCharges).flatMap(_.lumpSumBenefitTakenInExcessOfLifetimeAllowance),
         pensionPaidAnotherWayQuestion = prior.pensionCharges.flatMap(_.pensionSavingsTaxCharges).map(_.benefitInExcessOfLifetimeAllowance.isDefined),
-        pensionPaidAnotherWay = prior.pensionCharges.flatMap(_.pensionSavingsTaxCharges).flatMap(_.benefitInExcessOfLifetimeAllowance),
+        pensionPaidAnotherWay = prior.pensionCharges.flatMap(_.pensionSavingsTaxCharges).flatMap(_.benefitInExcessOfLifetimeAllowance).getOrElse(LifetimeAllowance(None, None)),
```
Comment (erasmos-hmrc) [inline app/models/pension/AllPensionsData.scala:78]: Switch to merely LifetimeAllowance(), once both values default to 'None'.

File: app/models/pension/charges/PensionLifetimeAllowancesViewModel.scala [hunk 6]
```diff
@@ -29,7 +29,7 @@ case class PensionLifetimeAllowancesViewModel(
                                                pensionAsLumpSumQuestion: Option[Boolean] = None,
                                                pensionAsLumpSum: Option[LifetimeAllowance] = None,
                                                pensionPaidAnotherWayQuestion: Option[Boolean] = None,
-                                               pensionPaidAnotherWay: Option[LifetimeAllowance] = None,
+                                               pensionPaidAnotherWay: LifetimeAllowance = LifetimeAllowance(None, None),
```
Comment (erasmos-hmrc) [inline app/models/pension/charges/PensionLifetimeAllowancesViewModel.scala:32]: Switch to merely LifetimeAllowance(), once both values default to 'None'.

File: app/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountController.scala [hunk 7]
```diff
@@ -21,7 +21,7 @@ import controllers.pensions.routes.PensionsSummaryController
 import controllers.pensions.lifetimeAllowance.routes.PensionTakenAnotherWayAmountController
 import controllers.predicates.AuthorisedAction
 import controllers.predicates.TaxYearAction.taxYearAction
-import forms.{FormUtils, TupleAmountForm}
+import forms.{FormUtils, OptionalTupleAmountForm, TupleAmountForm}
```
Comment (erasmos-hmrc) [inline app/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountController.scala:24]: TupleAmountForm is unused.

File: app/controllers/pensions/lifetimeAllowance/LifeTimeAllowanceAnotherWayController.scala [hunk 8]
```diff
@@ -74,7 +75,7 @@ class LifeTimeAllowanceAnotherWayController @Inject()(implicit val cc: MessagesC
               pensionsCYAModel.copy(
                 pensionLifetimeAllowances = viewModel.copy(
                   pensionPaidAnotherWayQuestion = Some(yesNo),
-                  pensionPaidAnotherWay = if (yesNo) viewModel.pensionPaidAnotherWay else None)
+                  pensionPaidAnotherWay = if (yesNo) charges.LifetimeAllowance(viewModel.pensionPaidAnotherWay.amount, viewModel.pensionPaidAnotherWay.taxPaid) else LifetimeAllowance(None, None))
```
Comment (erasmos-hmrc) [inline app/controllers/pensions/lifetimeAllowance/LifeTimeAllowanceAnotherWayController.scala:78]: Switch to merely LifetimeAllowance(), once both values default to 'None'.

File: app/controllers/pensions/lifetimeAllowance/AboveAnnualLifeTimeAllowanceController.scala [hunk 9]
```diff
@@ -76,7 +76,7 @@ class AboveAnnualLifeTimeAllowanceController @Inject()(implicit val cc: Messages
                   pensionAsLumpSumQuestion = if (yesNo) viewModel.pensionAsLumpSumQuestion else None,
                   pensionAsLumpSum = if (yesNo) viewModel.pensionAsLumpSum else None,
                   pensionPaidAnotherWayQuestion = if (yesNo) viewModel.pensionPaidAnotherWayQuestion else None,
-                  pensionPaidAnotherWay = if (yesNo) viewModel.pensionPaidAnotherWay else None)
+                  pensionPaidAnotherWay = if (yesNo) LifetimeAllowance(viewModel.pensionPaidAnotherWay.amount, viewModel.pensionPaidAnotherWay.taxPaid) else LifetimeAllowance(None, None))
```
Comment (erasmos-hmrc) [inline app/controllers/pensions/lifetimeAllowance/AboveAnnualLifeTimeAllowanceController.scala:79]: Switch to merely LifetimeAllowance(), once both values default to 'None'.

**Review by erasmos-hmrc — CHANGES_REQUESTED** [review]
Mostly suggestions except for the missing test scenario.

File: app/models/pension/charges/LifetimeAllowance.scala (hunk 1 above)
Comment (srinivaspach) [reply app/models/pension/charges/LifetimeAllowance.scala:27]: Set the default values to None

File: app/models/pension/charges/LifetimeAllowance.scala (hunk 2 above)
Comment (srinivaspach) [reply app/models/pension/charges/LifetimeAllowance.scala:43]: Set the default values to None

File: it/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountControllerISpec.scala (hunk 3 above)
Comment (srinivaspach) [reply it/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountControllerISpec.scala:302]: Yes I have now added it.

File: app/models/pension/AllPensionsData.scala (hunk 5 above)
Comment (srinivaspach) [reply app/models/pension/AllPensionsData.scala:78]: Done. Change implemented

File: app/models/pension/charges/PensionLifetimeAllowancesViewModel.scala (hunk 6 above)
Comment (srinivaspach) [reply app/models/pension/charges/PensionLifetimeAllowancesViewModel.scala:32]: Change implemented

File: app/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountController.scala (hunk 7 above)
Comment (srinivaspach) [reply app/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountController.scala:24]: TupleAmountForm removed from imports

File: app/controllers/pensions/lifetimeAllowance/LifeTimeAllowanceAnotherWayController.scala (hunk 8 above)
Comment (srinivaspach) [reply app/controllers/pensions/lifetimeAllowance/LifeTimeAllowanceAnotherWayController.scala:78]: Change implemented

File: app/controllers/pensions/lifetimeAllowance/AboveAnnualLifeTimeAllowanceController.scala (hunk 9 above)
Comment (srinivaspach) [reply app/controllers/pensions/lifetimeAllowance/AboveAnnualLifeTimeAllowanceController.scala:79]: Change implemented

File: app/models/pension/charges/LifetimeAllowance.scala (hunk 1 above)
Comment (erasmos-hmrc) [reply app/models/pension/charges/LifetimeAllowance.scala:27]: Confirmed.

File: app/models/pension/charges/LifetimeAllowance.scala (hunk 2 above)
Comment (erasmos-hmrc) [reply app/models/pension/charges/LifetimeAllowance.scala]: Confirmed.

File: it/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountControllerISpec.scala (hunk 3 above)
Comment (erasmos-hmrc) [reply it/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountControllerISpec.scala]: Confirmed.

File: app/models/pension/AllPensionsData.scala (hunk 5 above)
Comment (erasmos-hmrc) [reply app/models/pension/AllPensionsData.scala]: Confirmed.

File: app/models/pension/charges/PensionLifetimeAllowancesViewModel.scala (hunk 6 above)
Comment (erasmos-hmrc) [reply app/models/pension/charges/PensionLifetimeAllowancesViewModel.scala]: Confirmed.

File: app/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountController.scala (hunk 7 above)
Comment (erasmos-hmrc) [reply app/controllers/pensions/lifetimeAllowance/PensionTakenAnotherWayAmountController.scala]: Confirmed.

File: app/controllers/pensions/lifetimeAllowance/AboveAnnualLifeTimeAllowanceController.scala (hunk 9 above)
Comment (erasmos-hmrc) [reply app/controllers/pensions/lifetimeAllowance/AboveAnnualLifeTimeAllowanceController.scala]: Confirmed.

**Review by erasmos-hmrc — APPROVED** [review]
LGTM

**Pull request merged**